### DISCUSSION TITLE
Native function signature

### DISCUSSION THREAD

**Discussion by AlixANNERAUD** [bytecodealliance/wasm-micro-runtime#3479] [Q&A]
Hello, I was wondering how WAMR determines the signature of native functions since the signature strings of these functions were recently removed from registration. Indeed, I noticed that WAMR is able to determine, for example, whether the function requires an execution environment (`wasm_exec_env_t`) parameter or not.

**Comment by loganek** [comment]
Could you clarify what do you mean by signature strings being removed? They're still required: https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/lib_export.h#L23

AFAIK `wasm_exec_env_t` is always required for the native function; do you have any example of working code that doesn't require it?

**Reply by AlixANNERAUD** [reply]
Alright, I was confused by the string signature passage from `register_host_function` to `NativeSymbol`, but why the function signature is absent in [wamr-rust-sdk](https://github.com/bytecodealliance/wamr-rust-sdk/blob/269b84460f7b7ef322125c50d3d4454cd0d5139d/src/host_function.rs#L56) (it is always set to `null`).

Well, I had experimented with and without (in Rust) `wasm_exec_env_t` and strangely it still worked. I am trying to find an example for you

**Reply by lum1n0us** [reply]
If it is about `wasm_exec_env_t`, [it](https://github.com/bytecodealliance/wamr-rust-sdk/blob/main/src/function.rs#L88C1-L93C11) may answer your concern

**Reply by AlixANNERAUD** [reply]
> If it is about `wasm_exec_env_t`, [it](https://github.com/bytecodealliance/wamr-rust-sdk/blob/main/src/function.rs#L88C1-L93C11) may answer your concern

In this case, a function exported by WASM is being called from native. I may have explained myself poorly earlier; what I'm actually interested in is the reverse scenario—when native host functions (registered in the runtime) are called from WASM.

**Reply by AlixANNERAUD** [reply]
> Could you clarify what do you mean by signature strings being removed? They're still required: https://github.com/bytecodealliance/wasm-micro-runtime/blob/main/core/iwasm/include/lib_export.h#L23
> 
> AFAIK `wasm_exec_env_t` is always required for the native function; do you have any example of working code that doesn't require it?

[Here](https://github.com/bytecodealliance/wamr-rust-sdk/blob/269b84460f7b7ef322125c50d3d4454cd0d5139d/src/lib.rs#L116) is an example of missing `wasm_exec_env_t`.

**Reply by AlixANNERAUD** [reply]
This is probably not the best example I've found since this function has no parameters. However, I encountered this behavior and I don't understand how it could have occurred because, without wasm_exec_env_t, the other parameters (if they are passed blindly) are supposed to overlap each other. But that didn't happen.

**Reply by lum1n0us** [reply]
If there is a string in wasm(not sure will it ends with `\0` if the guest language is Rust, But if with C，there is one '\0'), its wasm getter(if it exists) will return its starting index(since the string is saved in linear memory) as an `i32`. 

After host getting the `i32`, it requires a address space translation for directly accessing or use API(like `wasm_memory_data()`) as a helper, and a move action from wasm world to host world which isn't because host functions can access linear memory. The real reason is rust using different string encoding with c.

We tend to pack all above details in rust-sdk. So rust-sdk users can have a rust string filled with the content of the string in wasm.

**Reply by AlixANNERAUD** [reply]
Alright, but my question now is how to pass strings and mutable buffers concretely with rust-sdk?

**Reply by lum1n0us** [reply]
I am working on sth. about passing strings between host and wasm. If you are interested, please follow https://github.com/bytecodealliance/wamr-rust-sdk/pull/39. It is still in progress but is enough to explain above idea.

**Reply by AlixANNERAUD** [reply]
This is exactly what I'm looking for. I can't wait for this to be released.
